src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Csrf;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+
+/**
+ * This CSRF token manager uses a combination of cookie and headers to validate non-persistent tokens.
+ *
+ * Double-Submit Validation: A JavaScript snippet on the client side is responsible for performing the
+ * double-submission. If the double-submit information is missing, we fall back to using the Origin or
+ * Referer headers.
+ *
+ * Fallback Scenarios: If neither double-submit nor Origin/Referer headers are available, it typically
+ * indicates that JavaScript is disabled on the client side (unless the JavaScript snippet was not
+ * properly implemented), or that the Origin header was not sent.
+ *
+ * By default, requests lacking both double-submit and origin information are deemed insecure.
+ *
+ * Security Consistency: When a session is found, a behavioral check is added to ensure that the
+ * validation method does not downgrade from double-submit to origin checks, or from origin checks to
+ * the accept fallback. This prevents attackers from exploiting potentially less secure validation
+ * methods once a more secure method has been confirmed as functional.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class DoubleSubmitCsrfTokenManager implements CsrfTokenManagerInterface
+{
+    public const HEADER_NAME = 'x-csrf-token';
+
+    public function __construct(
+        private RequestStack $requestStack,
+        private ?LoggerInterface $logger = null,
+        private string $headerName = self::HEADER_NAME,
+        private bool $acceptAsFallback = false,
```

Comment:
The name "acceptAsFallback" confuses me a bit... 

i would imagine `acceptAsFallback([$this, $that])` .. 

If i compare to an equivalent_-ish_ name from the Voters `low_if_all_abstain`"...  maybe here  "allow_if_no_failure" "ignore_missing_origin"  or something like that ?

Reply:
naming... this one took me too long already :D 
Let me remove that option, that's the safer way.
People that don't have neither Origin nor Referer should have JS instead.